# Capitulo 5: Aprendizaje automatico 1

**Contenido del capitulo**
- Unidad 1. Analisis de datos basado en el aprendizaje automatico.
- Unidad 2. Aplicacion del modelo de aprendizaje supervisado para la prediccion numerica.
- Unidad 3. Aplicacion del modelo de aprendizaje supervisado para la clasificacion.
- Unidad 4. Arbol de decisiones
- Unidad 5. Algoritmo Naive Bayes
- Unidad 6. Algoritmo KNN
- Unidad 7. Algoritmo SVM
- Unidad 8. Algoritmos de conjunto

# Unidad 1. Análisis de datos basado en el aprendizaje automático

## 1.2. Biblioteca Python scikit-learn para el aprendizaje automático

In [33]:
# El modulo sklearn.datasets incluye utilizadas para cargar conjuntos de datos,
# incluyendo metodos para cargar y obtener conjuntos de datos de referencia. 
# Tambien incluye algunos generadores de datos artificiales
import pandas as pd
import numpy as np
from sklearn.datasets import  load_breast_cancer
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.tree import DecisionTreeClassifier


In [34]:
type(load_breast_cancer())

sklearn.utils._bunch.Bunch

In [35]:
data = load_breast_cancer()

In [36]:
pd.DataFrame(data['data']) # Estos datos se convierten en x (variable independiente)

0      1       2       3        4        5        6        7       8   \
0    17.99  10.38  122.80  1001.0  0.11840  0.27760  0.30010  0.14710  0.2419   
1    20.57  17.77  132.90  1326.0  0.08474  0.07864  0.08690  0.07017  0.1812   
2    19.69  21.25  130.00  1203.0  0.10960  0.15990  0.19740  0.12790  0.2069   
3    11.42  20.38   77.58   386.1  0.14250  0.28390  0.24140  0.10520  0.2597   
4    20.29  14.34  135.10  1297.0  0.10030  0.13280  0.19800  0.10430  0.1809   
..     ...    ...     ...     ...      ...      ...      ...      ...     ...   
564  21.56  22.39  142.00  1479.0  0.11100  0.11590  0.24390  0.13890  0.1726   
565  20.13  28.25  131.20  1261.0  0.09780  0.10340  0.14400  0.09791  0.1752   
566  16.60  28.08  108.30   858.1  0.08455  0.10230  0.09251  0.05302  0.1590   
567  20.60  29.33  140.10  1265.0  0.11780  0.27700  0.35140  0.15200  0.2397   
568   7.76  24.54   47.92   181.0  0.05263  0.04362  0.00000  0.00000  0.1587   

          9   ...      20     21      22      23       24       25      26  \
0    0.07871  ...  25.380  17.33  184.60  2019.0  0.16220  0.66560  0.7119   
1    0.05667  ...  24.990  23.41  158.80  1956.0  0.12380  0.18660  0.2416   
2    0.05999  ...  23.570  25.53  152.50  1709.0  0.14440  0.42450  0.4504   
3    0.09744  ...  14.910  26.50   98.87   567.7  0.20980  0.86630  0.6869   
4    0.05883  ...  22.540  16.67  152.20  1575.0  0.13740  0.20500  0.4000   
..       ...  ...     ...    ...     ...     ...      ...      ...     ...   
564  0.05623  ...  25.450  26.40  166.10  2027.0  0.14100  0.21130  0.4107   
565  0.05533  ...  23.690  38.25  155.00  1731.0  0.11660  0.19220  0.3215   
566  0.05648  ...  18.980  34.12  126.70  1124.0  0.11390  0.30940  0.3403   
567  0.07016  ...  25.740  39.42  184.60  1821.0  0.16500  0.86810  0.9387   
568  0.05884  ...   9.456  30.37   59.16   268.6  0.08996  0.06444  0.0000   

         27      28       29  
0    0.2654  0.4601  0.11890  
1    0.1860  0.2750  0.08902  
2    0.2430  0.3613  0.08758  
3    0.2575  0.6638  0.17300  
4    0.1625  0.2364  0.07678  
..      ...     ...      ...  
564  0.2216  0.2060  0.07115  
565  0.1628  0.2572  0.06637  
566  0.1418  0.2218  0.07820  
567  0.2650  0.4087  0.12400  
568  0.0000  0.2871  0.07039  

[569 rows x 30 columns]

In [37]:
pd.DataFrame(data['target'])

0
0    0
1    0
2    0
3    0
4    0
..  ..
564  0
565  0
566  0
567  0
568  1

[569 rows x 1 columns]

In [38]:
data.DESCR # proporciona detalles sobre los datos

'.. _breast_cancer_dataset:\n\nBreast cancer wisconsin (diagnostic) dataset\n--------------------------------------------\n\n**Data Set Characteristics:**\n\n:Number of Instances: 569\n\n:Number of Attributes: 30 numeric, predictive attributes and the class\n\n:Attribute Information:\n    - radius (mean of distances from center to points on the perimeter)\n    - texture (standard deviation of gray-scale values)\n    - perimeter\n    - area\n    - smoothness (local variation in radius lengths)\n    - compactness (perimeter^2 / area - 1.0)\n    - concavity (severity of concave portions of the contour)\n    - concave points (number of concave portions of the contour)\n    - symmetry\n    - fractal dimension ("coastline approximation" - 1)\n\n    The mean, standard error, and "worst" or largest (mean of the three\n    worst/largest values) of these features were computed for each image,\n    resulting in 30 features.  For instance, field 0 is Mean Radius, field\n    10 is Radius SE, field 

In [39]:
help(train_test_split)

Help on function train_test_split in module sklearn.model_selection._split:

train_test_split(
    *arrays,
    test_size=None,
    train_size=None,
    random_state=None,
    shuffle=True,
    stratify=None
)
    Split arrays or matrices into random train and test subsets.

    Quick utility that wraps input validation,
    ``next(ShuffleSplit().split(X, y))``, and application to input data
    into a single call for splitting (and optionally subsampling) data into a
    one-liner.

    Read more in the :ref:`User Guide <cross_validation>`.

    Parameters
    ----------
    *arrays : sequence of indexables with same length / shape[0]
        Allowed inputs are lists, numpy arrays, scipy-sparse
        matrices or pandas dataframes.

    test_size : float or int, default=None
        If float, should be between 0.0 and 1.0 and represent the proportion
        of the dataset to include in the test split. If int, represents the
        absolute number of test samples. If None, the value

In [40]:
len(data.data)

569

In [41]:
train_test_split

<function sklearn.model_selection._split.train_test_split(*arrays, test_size=None, train_size=None, random_state=None, shuffle=True, stratify=None)>

In [42]:
X_train, X_test, y_train, y_test=train_test_split(data.data,data.target)

In [43]:
len(X_train)

426

In [44]:
569*0.75

426.75

In [45]:
569*0.25

142.25

In [46]:
len(X_test)

143

In [47]:
data = load_breast_cancer()
X_train, X_test, y_train, y_test=train_test_split(data.data,data.target, random_state=42)

model = DecisionTreeClassifier(criterion='entropy')
model

DecisionTreeClassifier(criterion='entropy')

In [48]:
model.fit(X_train, y_train)
y_pred=model.predict(X_test)
y_pred

array([1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1,
       0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1,
       0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 1, 1,
       0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0,
       1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1])

In [49]:
pred = model.predict(X_test)

In [50]:
pred_2d=pred.reshape(len(pred),1)
pred_2d

array([[1],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [0],
       [0],
       [1],
       [1],
       [0],
       [1],
       [1],
       [1],
       [0],
       [1],
       [1],
       [1],
       [0],
       [1],
       [1],
       [0],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [0],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [0],
       [1],
       [0],
       [1],
       [1],
       [0],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [0],
       [0],
       [1],
       [1],
       [1],
       [1],
       [1],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [1],
       [1],
       [1],
       [0],
       [0],
       [1],
       [1],
       [0],
       [0],
       [1],
       [0],
       [1],
       [1],
       [1],
       [1],
       [1],
       [1],
       [0],
       [1],
       [1],
    

In [51]:
y_test_2d=y_test.reshape(len(pred),1)
df1 = pd.DataFrame(pred_2d)
df2 = pd.DataFrame(y_test_2d)

In [52]:
# Anteriormente creamos dos df cada uno de dimension 143 x 1. Por lo que
# Vamos a concatenarlos para crear nuestra matriz de los predichos vs los observados
df_concat = pd.concat([df1,df2],axis=1) # matriz 143 x 2
# Redefinimos los nombres de las columnas
df_concat.columns=["predicho","real"]

In [53]:
# Ahora queremos comparar cuando estos son iguales 
df_concat[df_concat['predicho'] == df_concat['real']]

predicho  real
0           1     1
1           0     0
2           0     0
3           1     1
4           1     1
..        ...   ...
138         1     1
139         0     0
140         1     1
141         0     0
142         1     1

[137 rows x 2 columns]

In [54]:
# Queremos determinar un porcentaje de exitos (exito = sumatoria (real == predicho))
len(df_concat[df_concat['predicho'] == df_concat['real']])

137

In [55]:
133/143 # Porcertaje muy malo. Se recomienda estandarizar para solucionar el problema

0.9300699300699301

In [56]:
scaler = StandardScaler()

In [57]:
X_train

array([[1.289e+01, 1.312e+01, 8.189e+01, ..., 5.366e-02, 2.309e-01,
        6.915e-02],
       [1.340e+01, 2.052e+01, 8.864e+01, ..., 2.051e-01, 3.585e-01,
        1.109e-01],
       [1.296e+01, 1.829e+01, 8.418e+01, ..., 6.608e-02, 3.207e-01,
        7.247e-02],
       ...,
       [1.429e+01, 1.682e+01, 9.030e+01, ..., 3.333e-02, 2.458e-01,
        6.120e-02],
       [1.398e+01, 1.962e+01, 9.112e+01, ..., 1.827e-01, 3.179e-01,
        1.055e-01],
       [1.218e+01, 2.052e+01, 7.722e+01, ..., 7.431e-02, 2.694e-01,
        6.878e-02]], shape=(426, 30))

In [58]:
df_scale_previous=pd.DataFrame(data['data']) # marco de datos antes de la normalizacion
scaler.fit(X_train)

StandardScaler()

In [59]:
X_train=scaler.transform(X_train)
df_scale_after=pd.DataFrame(X_train)

In [60]:
df_scale_previous

0      1       2       3        4        5        6        7       8   \
0    17.99  10.38  122.80  1001.0  0.11840  0.27760  0.30010  0.14710  0.2419   
1    20.57  17.77  132.90  1326.0  0.08474  0.07864  0.08690  0.07017  0.1812   
2    19.69  21.25  130.00  1203.0  0.10960  0.15990  0.19740  0.12790  0.2069   
3    11.42  20.38   77.58   386.1  0.14250  0.28390  0.24140  0.10520  0.2597   
4    20.29  14.34  135.10  1297.0  0.10030  0.13280  0.19800  0.10430  0.1809   
..     ...    ...     ...     ...      ...      ...      ...      ...     ...   
564  21.56  22.39  142.00  1479.0  0.11100  0.11590  0.24390  0.13890  0.1726   
565  20.13  28.25  131.20  1261.0  0.09780  0.10340  0.14400  0.09791  0.1752   
566  16.60  28.08  108.30   858.1  0.08455  0.10230  0.09251  0.05302  0.1590   
567  20.60  29.33  140.10  1265.0  0.11780  0.27700  0.35140  0.15200  0.2397   
568   7.76  24.54   47.92   181.0  0.05263  0.04362  0.00000  0.00000  0.1587   

          9   ...      20     21      22      23       24       25      26  \
0    0.07871  ...  25.380  17.33  184.60  2019.0  0.16220  0.66560  0.7119   
1    0.05667  ...  24.990  23.41  158.80  1956.0  0.12380  0.18660  0.2416   
2    0.05999  ...  23.570  25.53  152.50  1709.0  0.14440  0.42450  0.4504   
3    0.09744  ...  14.910  26.50   98.87   567.7  0.20980  0.86630  0.6869   
4    0.05883  ...  22.540  16.67  152.20  1575.0  0.13740  0.20500  0.4000   
..       ...  ...     ...    ...     ...     ...      ...      ...     ...   
564  0.05623  ...  25.450  26.40  166.10  2027.0  0.14100  0.21130  0.4107   
565  0.05533  ...  23.690  38.25  155.00  1731.0  0.11660  0.19220  0.3215   
566  0.05648  ...  18.980  34.12  126.70  1124.0  0.11390  0.30940  0.3403   
567  0.07016  ...  25.740  39.42  184.60  1821.0  0.16500  0.86810  0.9387   
568  0.05884  ...   9.456  30.37   59.16   268.6  0.08996  0.06444  0.0000   

         27      28       29  
0    0.2654  0.4601  0.11890  
1    0.1860  0.2750  0.08902  
2    0.2430  0.3613  0.08758  
3    0.2575  0.6638  0.17300  
4    0.1625  0.2364  0.07678  
..      ...     ...      ...  
564  0.2216  0.2060  0.07115  
565  0.1628  0.2572  0.06637  
566  0.1418  0.2218  0.07820  
567  0.2650  0.4087  0.12400  
568  0.0000  0.2871  0.07039  

[569 rows x 30 columns]

In [61]:
df_scale_after

0         1         2         3         4         5         6   \
0   -0.349138 -1.438513 -0.411726 -0.390479 -1.863662 -1.268607 -0.826171   
1   -0.204687  0.312640 -0.133673 -0.275880  1.078073  0.863546  0.726314   
2   -0.329312 -0.215072 -0.317394 -0.364357 -1.579880 -0.457451 -0.597310   
3    1.027403  2.089824  1.046922  0.917584  0.316303  0.562037  1.048527   
4    1.828969  0.696001  1.763681  1.783821 -0.333674  0.628175  0.974660   
..        ...       ...       ...       ...       ...       ...       ...   
421 -1.482661 -1.078817 -1.363286 -1.154195  0.162946  0.984150 -0.017962   
422 -0.703187 -0.205607 -0.688544 -0.680910  1.336057 -0.015692 -0.213582   
423  0.047396 -0.562937 -0.065292 -0.062691 -2.240606 -1.473633 -1.021664   
424 -0.040408  0.099662 -0.031514 -0.155663  0.748426  0.209953  0.320044   
425 -0.550238  0.312640 -0.604098 -0.551143 -1.105476 -1.208500 -0.810506   

           7         8         9   ...        20        21        22  \
0   -0.952866 -1.729368 -0.941541  ... -0.542685 -1.655655 -0.589864   
1    0.898441  1.178770  1.474377  ...  0.041487  0.689899  0.194128   
2   -0.764588  0.275343 -0.501024  ... -0.435901 -0.148985 -0.320159   
3    0.930437 -0.325697 -0.477474  ...  1.113515  2.165006  1.165793   
4    1.265740 -0.131572 -1.713139  ...  1.471556  0.387568  1.556276   
..        ...       ...       ...  ...       ...       ...       ...   
421 -0.503063  0.379872  3.767006  ... -1.356547 -1.634059 -1.341770   
422 -0.340964  0.002822  0.359231  ... -0.643188  0.623452 -0.647074   
423 -1.097247 -1.090996 -1.225522  ... -0.272584 -0.806803 -0.376764   
424  0.446522 -0.489956  0.392478  ...  0.173397  0.879270  0.212290   
425 -0.794470 -0.228635 -0.808555  ... -0.601311  1.218146 -0.675225   

           23        24        25        26        27        28        29  
0   -0.525560 -1.510669 -0.891500 -0.750217 -0.916711 -0.925086 -0.808411  
1   -0.051934  1.129415  0.923942  1.222217  1.436560  1.149559  1.569111  
2   -0.446032 -1.634396 -0.106752 -0.539891 -0.723713  0.534970 -0.619348  
3    0.997696  0.383604  0.860948  1.872819  1.310691  0.152884  0.421636  
4    1.385595 -0.577759  0.296680  0.595768  1.232995  0.050452 -1.406351  
..        ...       ...       ...       ...       ...       ...       ...  
421 -1.043821 -0.452739 -0.027674 -0.625430 -1.006838 -1.014510  1.426745  
422 -0.623509  1.620874  0.122440  0.098634  0.142137 -0.117015  0.434165  
423 -0.334976 -1.962898 -1.322676 -1.152453 -1.232624 -0.682827 -1.261137  
424 -0.007830  1.297546  0.730938  0.700428  1.088480  0.489445  1.261600  
425 -0.577280 -0.814867 -1.066276 -0.770847 -0.595824 -0.299115 -0.829481  

[426 rows x 30 columns]

In [62]:
# Salida antes del pretratamiento
X_train

array([[-0.34913849, -1.43851335, -0.41172595, ..., -0.91671059,
        -0.92508585, -0.80841115],
       [-0.20468665,  0.31264011, -0.13367256, ...,  1.43655962,
         1.14955889,  1.56911143],
       [-0.32931176, -0.21507235, -0.31739376, ..., -0.7237126 ,
         0.53496977, -0.61934827],
       ...,
       [ 0.04739597, -0.56293662, -0.06529202, ..., -1.23262438,
        -0.68282718, -1.261137  ],
       [-0.04040808,  0.09966199, -0.03151368, ...,  1.08847951,
         0.48944465,  1.26159953],
       [-0.5502381 ,  0.31264011, -0.6040977 , ..., -0.59582424,
        -0.29911546, -0.82948141]], shape=(426, 30))

In [63]:
# Preprocesamiento - Aplicar la escala
scaler = StandardScaler()
scaler.fit(X_train)
X_train = scaler.transform(X_train)

In [64]:
# Comporbacion  del resultado tras el preprocesamiento
X_train

array([[-0.34913849, -1.43851335, -0.41172595, ..., -0.91671059,
        -0.92508585, -0.80841115],
       [-0.20468665,  0.31264011, -0.13367256, ...,  1.43655962,
         1.14955889,  1.56911143],
       [-0.32931176, -0.21507235, -0.31739376, ..., -0.7237126 ,
         0.53496977, -0.61934827],
       ...,
       [ 0.04739597, -0.56293662, -0.06529202, ..., -1.23262438,
        -0.68282718, -1.261137  ],
       [-0.04040808,  0.09966199, -0.03151368, ...,  1.08847951,
         0.48944465,  1.26159953],
       [-0.5502381 ,  0.31264011, -0.6040977 , ..., -0.59582424,
        -0.29911546, -0.82948141]], shape=(426, 30))

In [65]:
X_test # Antes

array([[1.247e+01, 1.860e+01, 8.109e+01, ..., 1.015e-01, 3.014e-01,
        8.750e-02],
       [1.894e+01, 2.131e+01, 1.236e+02, ..., 1.789e-01, 2.551e-01,
        6.589e-02],
       [1.546e+01, 1.948e+01, 1.017e+02, ..., 1.514e-01, 2.837e-01,
        8.019e-02],
       ...,
       [1.104e+01, 1.683e+01, 7.092e+01, ..., 7.431e-02, 2.998e-01,
        7.881e-02],
       [1.981e+01, 2.215e+01, 1.300e+02, ..., 2.388e-01, 2.768e-01,
        7.615e-02],
       [1.026e+01, 1.222e+01, 6.575e+01, ..., 6.696e-02, 2.937e-01,
        7.722e-02]], shape=(143, 30))

In [66]:
X_test = scaler.fit_transform(X_test)

X_test # Despues

array([[-0.47853809, -0.21371678, -0.45835473, ..., -0.27372978,
         0.07289581,  0.09152022],
       [ 1.37424404,  0.38950001,  1.29239853, ...,  0.84579664,
        -0.67110896, -1.02766663],
       [ 0.37769352, -0.01783827,  0.3904579 , ...,  0.44803208,
        -0.21152934, -0.28706636],
       ...,
       [-0.88804017, -0.60769971, -0.87720113, ..., -0.66701045,
         0.04718507, -0.35853687],
       [ 1.62338166,  0.57647495,  1.55597935, ...,  1.71220016,
        -0.32240694, -0.49629888],
       [-1.11140494, -1.63383602, -1.09012501, ..., -0.77332207,
        -0.05083716, -0.44088334]], shape=(143, 30))

# 1.4. Preprocesamiento de datos para hacer un buen conjunto de datos de entrenamiento

## Procesamiento de valores perdidos

In [67]:
import pandas as pd
import numpy as np
from sklearn.datasets import load_iris
iris= load_iris()
iris['data']

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

### Procesamiento de valores perdidos

1. Ignorar el registro (fila)
2. Insertar el valor que falta (imputacion)
3. Entrada manual (se encarga un responsable o experto)

In [68]:
data = [[1.0,2.0,3.0,4.0],[5.0,6.0,np.nan,8.0],[10.0,11.0,12.0,np.nan]]

In [69]:
np.array(data)

array([[ 1.,  2.,  3.,  4.],
       [ 5.,  6., nan,  8.],
       [10., 11., 12., nan]])

In [70]:
pd.DataFrame(np.array(data), columns=['A','B','C','D'])

A     B     C    D
0   1.0   2.0   3.0  4.0
1   5.0   6.0   NaN  8.0
2  10.0  11.0  12.0  NaN

In [71]:
df = pd.DataFrame(np.array(data), columns=['A','B','C','D'])

**isnull():** devuelve un marco de datos con un booleano que muestra si la celda tiene un valor numerico (Falso) o se omite con un valor numerico (Verdadero).

**isnull().sum():** devuelve el numero de omisiones

In [72]:
df.isnull().sum()

A    0
B    0
C    1
D    1
dtype: int64

### Primer caso: Eliminar las filas o columnas con valores nulos

In [73]:
df.dropna() # axis=0 es el valor por defecto, por lo que la fila con el valor NaN se elimina

A    B    C    D
0  1.0  2.0  3.0  4.0

In [74]:
df

A     B     C    D
0   1.0   2.0   3.0  4.0
1   5.0   6.0   NaN  8.0
2  10.0  11.0  12.0  NaN

In [75]:
df.dropna(inplace=True) # inplace=True para que los cambios se guarden en el mismo df

In [76]:
df

A    B    C    D
0  1.0  2.0  3.0  4.0

In [77]:
df = pd.DataFrame(np.array(data), columns=['A','B','C','D'])
df

A     B     C    D
0   1.0   2.0   3.0  4.0
1   5.0   6.0   NaN  8.0
2  10.0  11.0  12.0  NaN

In [78]:
df.dropna(axis=1, inplace=True) # Elimina las columnas con valores nulos
df

A     B
0   1.0   2.0
1   5.0   6.0
2  10.0  11.0

In [79]:
df = pd.DataFrame(np.array(data), columns=['A','B','C','D'])
df

A     B     C    D
0   1.0   2.0   3.0  4.0
1   5.0   6.0   NaN  8.0
2  10.0  11.0  12.0  NaN

Nota: **Si todas las filas tienen NaN, usa how='all' para borrar**

In [80]:
df.loc[3]=[np.nan,np.nan,np.nan,np.nan]
df

A     B     C    D
0   1.0   2.0   3.0  4.0
1   5.0   6.0   NaN  8.0
2  10.0  11.0  12.0  NaN
3   NaN   NaN   NaN  NaN

In [81]:
df.dropna(how='all', inplace=True) # Elimina las filas con todos los valores nulos
df

A     B     C    D
0   1.0   2.0   3.0  4.0
1   5.0   6.0   NaN  8.0
2  10.0  11.0  12.0  NaN

In [82]:
df.dropna(subset=['C'], inplace=True) # Elimina las filas con valores nulos en la columna C
df

A     B     C    D
0   1.0   2.0   3.0  4.0
2  10.0  11.0  12.0  NaN

### Segundo caso: Imputación

In [83]:
from sklearn.impute import SimpleImputer

In [84]:
df = pd.DataFrame(np.array(data), columns=['A','B','C','D'])
impt=SimpleImputer(missing_values=np.nan, strategy='mean')

In [85]:
df.values 

array([[ 1.,  2.,  3.,  4.],
       [ 5.,  6., nan,  8.],
       [10., 11., 12., nan]])

In [86]:
imputed = impt.fit(df.values) # Imputar con el uso de df.values

In [87]:
imputed.transform(df.values)

array([[ 1. ,  2. ,  3. ,  4. ],
       [ 5. ,  6. ,  7.5,  8. ],
       [10. , 11. , 12. ,  6. ]])

In [88]:
(3+12)/2 # Comprueba que es la media de la columna

7.5

La clase SimpleImputer es una clase transformadora de scikit-learn que se utiliza para la conversión de datos. Los dos métodos principales para el estimador son el ajuste y la transformación. 

Utilice el método de ajuste para aprender el parámetro del modelo en los datos de entrenamiento.

Utilice el método de transformación para convertir los datos en el parámetro aprendido.

La matriz de datos para la conversión debe ser igual al número de características de datos utilizados en el aprendizaje del modelo.

## Procesamiento de datos categóricos

In [89]:
# El tamaño esta ordenado , pero el color no lo esta.
# Por lo tanto, el tamaño se clasifica como escala ordinal mientras que el color es escala nominal
df = pd.DataFrame([['green','M', 10.1, 'class2'],
                  ['red','L', 13.5, 'class1'],
                  ['blue','XL', 15.3, 'class2']])
df

0   1     2       3
0  green   M  10.1  class2
1    red   L  13.5  class1
2   blue  XL  15.3  class2

**Nota:** *Convertir datos ordenados en valores numericos. La razon por la que se cambian los datos de texto en datos numericos es para que un ordenador pueda procesar y hacer operaciones aritmeticas*.

In [90]:
size_mapping = {'XL':3, 'L':2, 'M':1}
df.columns = ['color','size','price','classlabel']
df

color size  price classlabel
0  green    M   10.1     class2
1    red    L   13.5     class1
2   blue   XL   15.3     class2

In [91]:
df['size']

0     M
1     L
2    XL
Name: size, dtype: object

In [92]:
df['size'] = df['size'].map(size_mapping) # Mapea los valores de la columna size con el mapeo ya antes mencionado

In [93]:
df['size']

0    1
1    2
2    3
Name: size, dtype: int64

In [94]:
df

color  size  price classlabel
0  green     1   10.1     class2
1    red     2   13.5     class1
2   blue     3   15.3     class2

## Codificacion de classlabel

Crear un mapeo para convertir la etiqueta de la clase de cadenas a enteros.

In [95]:
df['classlabel']

0    class2
1    class1
2    class2
Name: classlabel, dtype: object

In [96]:
df['classlabel'].value_counts()

classlabel
class2    2
class1    1
Name: count, dtype: int64

In [97]:
np.unique(df['classlabel']) # Descubre los valores unicos de la columna classlabel y los ordena

array(['class1', 'class2'], dtype=object)

enumerate() crea un objeto con un indice

In [98]:
enumerate(np.unique(df['classlabel'])) # Enumera los valores unicos de la columna classlabel

In [99]:
for idx, lable in enumerate(np.unique(df['classlabel'])):
    print(idx,lable)

0 class1
1 class2


In [100]:
{label:idx for idx, label in enumerate(np.unique(df['classlabel']))}

{'class1': 0, 'class2': 1}

In [101]:
class_mapping = {label:idx for idx, label in enumerate(np.unique(df['classlabel']))}

In [102]:
df['classlabel'] = df['classlabel'].map(class_mapping) # Cambiar la etiqueta de la clase de cadenas a enteros
df

color  size  price  classlabel
0  green     1   10.1           1
1    red     2   13.5           0
2   blue     3   15.3           1

Dado que el metodo anterior es bastante inconveniente, scikit-learn soporta LabelEncoder para una facil conversion

In [103]:
df = pd.DataFrame([['green','M', 10.1, 'class2'],
                    ['red','L', 13.5, 'class1'],
                    ['blue','XL', 15.3, 'class2']])
df.columns = ['color','size','price','classlabel']
df

color size  price classlabel
0  green    M   10.1     class2
1    red    L   13.5     class1
2   blue   XL   15.3     class2

In [104]:
from sklearn.preprocessing import LabelEncoder
enc=LabelEncoder()
y=enc.fit_transform(df['classlabel'].values)
y

array([1, 0, 1])

## Aplicación de la codificación de una sola vez a una característica desordenada

Se crean variables ficticias en las que se mide si una caracteristica determinada esta presente o no, codificandola como ceros o como unos

In [105]:
iris.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

La codificacion se realiza con numeros enteros, por lo que hay que insertar el valor de la "especie" del iris

In [106]:
iris = pd.DataFrame(iris.target, columns=['species'])
iris

species
0          0
1          0
2          0
3          0
4          0
..       ...
145        2
146        2
147        2
148        2
149        2

[150 rows x 1 columns]

In [107]:
iris['species'].replace({0:'setosa',1:'versicolor',2:'virginica'}, inplace=True)
iris

C:\Users\Sebastian\AppData\Local\Temp\ipykernel_6624\35287107.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  iris['species'].replace({0:'setosa',1:'versicolor',2:'virginica'}, inplace=True)


species
0       setosa
1       setosa
2       setosa
3       setosa
4       setosa
..         ...
145  virginica
146  virginica
147  virginica
148  virginica
149  virginica

[150 rows x 1 columns]

get_dummies() permite convertir cada valor propio de las variables categoricas en una nueva variable ficticia.

In [108]:
pd.get_dummies(iris['species'])

setosa  versicolor  virginica
0      True       False      False
1      True       False      False
2      True       False      False
3      True       False      False
4      True       False      False
..      ...         ...        ...
145   False       False       True
146   False       False       True
147   False       False       True
148   False       False       True
149   False       False       True

[150 rows x 3 columns]

## OneHotEncoder

In [109]:
from sklearn.preprocessing import OneHotEncoder
onehot_encoder = OneHotEncoder()

In [110]:
len_iris=len(iris['species'])

In [111]:
np.array(iris['species'])

array(['setosa', 'setosa', 'setosa', 'setosa', 'setosa', 'setosa',
       'setosa', 'setosa', 'setosa', 'setosa', 'setosa', 'setosa',
       'setosa', 'setosa', 'setosa', 'setosa', 'setosa', 'setosa',
       'setosa', 'setosa', 'setosa', 'setosa', 'setosa', 'setosa',
       'setosa', 'setosa', 'setosa', 'setosa', 'setosa', 'setosa',
       'setosa', 'setosa', 'setosa', 'setosa', 'setosa', 'setosa',
       'setosa', 'setosa', 'setosa', 'setosa', 'setosa', 'setosa',
       'setosa', 'setosa', 'setosa', 'setosa', 'setosa', 'setosa',
       'setosa', 'setosa', 'versicolor', 'versicolor', 'versicolor',
       'versicolor', 'versicolor', 'versicolor', 'versicolor',
       'versicolor', 'versicolor', 'versicolor', 'versicolor',
       'versicolor', 'versicolor', 'versicolor', 'versicolor',
       'versicolor', 'versicolor', 'versicolor', 'versicolor',
       'versicolor', 'versicolor', 'versicolor', 'versicolor',
       'versicolor', 'versicolor', 'versicolor', 'versicolor',
       'versicolo

In [112]:
iris_species_2d=np.array(iris['species']).reshape(len_iris,1)

In [113]:
onehot_encoder.fit_transform(iris_species_2d)


<Compressed Sparse Row sparse matrix of dtype 'float64'
	with 150 stored elements and shape (150, 3)>

In [114]:
# Conversion a la matriz dispersa


## Uso de la retencion en la vida real que divide el conjunto de datos en conjunto de datos de entrenamiento y conjunto de prueba

In [115]:
import pandas as pd
import numpy as np
df_wine = pd.read_csv('https://archive.ics.uci.edu/'
                      'ml/machine-learning-databases/wine/wine.data',
                      header=None)
df_wine.columns = ['Class label', 'Alcohol', 'malic acid', 'Ash',
                   'Alcalinity of ash', 'Magnesium', 'Total phenols',
                   'Flavanoids', 'Nonflavanoid phenols', 'Proanthocyanins',
                   'Color intensity', 'Hue', 'OD280/OD315 of diluted wines',
                   'Proline']
print('Class labels', np.unique(df_wine['Class label']))
df_wine.head()


Class labels [1 2 3]


Class label  Alcohol  malic acid   Ash  Alcalinity of ash  Magnesium  \
0            1    14.23        1.71  2.43               15.6        127   
1            1    13.20        1.78  2.14               11.2        100   
2            1    13.16        2.36  2.67               18.6        101   
3            1    14.37        1.95  2.50               16.8        113   
4            1    13.24        2.59  2.87               21.0        118   

   Total phenols  Flavanoids  Nonflavanoid phenols  Proanthocyanins  \
0           2.80        3.06                  0.28             2.29   
1           2.65        2.76                  0.26             1.28   
2           2.80        3.24                  0.30             2.81   
3           3.85        3.49                  0.24             2.18   
4           2.80        2.69                  0.39             1.82   

   Color intensity   Hue  OD280/OD315 of diluted wines  Proline  
0             5.64  1.04                          3.92     1065  
1             4.38  1.05                          3.40     1050  
2             5.68  1.03                          3.17     1185  
3             7.80  0.86                          3.45     1480  
4             4.32  1.04                          2.93      735

In [116]:
from sklearn.model_selection import train_test_split
X,y = df_wine.iloc[:, 1:].values, df_wine.iloc[:,0].values

X_train, X_test, y_train, y_test = train_test_split(X,y, 
                                    test_size=0.3, # El 30% de la muestra se asigna a X_test y y_test
                                    random_state=0,
                                    stratify=y)

**Ordenacion de la escala entre caracteristicas (variables)**

**MinMaxScaler**: Esta clase se utiliza para realizar la normalización de características o escalamiento de datos. En este caso, MinMaxScaler escala cada característica a un rango específico (por defecto, entre 0 y 1).

In [117]:
from sklearn.preprocessing import MinMaxScaler

mms = MinMaxScaler()
X_train_norm = mms.fit_transform(X_train)
X_test_norm = mms.transform(X_test)

In [118]:
from sklearn.preprocessing import StandardScaler

stdsc = StandardScaler()
X_train_std = stdsc.fit_transform(X_train)
X_test_std = stdsc.transform(X_test)


In [119]:
from sklearn.preprocessing import RobustScaler
rbs = RobustScaler()
X_train_robust = rbs.fit_transform(X_train)
X_test_robust = rbs.transform(X_test)

**MaxAbsScaler**: Divide los datos en el valor absoluto maximo basado en cada caracteristica. Asi, el valor maximo de cada caracteristica se convierte en 1.

La caracteristica general cambia al rango  [-1,1]

In [120]:
from sklearn.preprocessing import MaxAbsScaler

mas = MaxAbsScaler()
X_train_maxabs = mas.fit_transform(X_train)
X_test_maxabs = mas.transform(X_test)

# 1.5. Practicando para encontrar un metodo optimo para resolver problemas con scikit-learn

### Entender los datos del iris

In [121]:
from sklearn.datasets import load_iris
data=load_iris()
data.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

In [122]:
print(data.DESCR)

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

:Number of Instances: 150 (50 in each of three classes)
:Number of Attributes: 4 numeric, predictive attributes and the class
:Attribute Information:
    - sepal length in cm
    - sepal width in cm
    - petal length in cm
    - petal width in cm
    - class:
            - Iris-Setosa
            - Iris-Versicolour
            - Iris-Virginica

:Summary Statistics:

============== ==== ==== ======= ===== ====================
                Min  Max   Mean    SD   Class Correlation
============== ==== ==== ======= ===== ====================
sepal length:   4.3  7.9   5.84   0.83    0.7826
sepal width:    2.0  4.4   3.05   0.43   -0.4194
petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
============== ==== ==== ======= ===== ====================

:Missing Attribute Values: None
:Class Distribution: 33.3% for each of 3 classes.
:Cr

### Preprocesamiento de datos del iris y EDA

In [123]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [124]:
iris= pd.DataFrame(data=data.data,columns=data.feature_names)
iris

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]

In [125]:
feature = pd.DataFrame(data.data,columns=data.feature_names)
data.target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [126]:
target=pd.DataFrame(data.target,columns=['species'])
target

species
0          0
1          0
2          0
3          0
4          0
..       ...
145        2
146        2
147        2
148        2
149        2

[150 rows x 1 columns]

In [127]:
# Fusionar la caracteristica y el objetivo
iris = pd.concat([feature,target],axis=1)
# Cambiar el nombre de la columna
iris.rename({"sepal length (cm)":"sepal_length","sepal width (cm)":"sepal_width",
             "petal length (cm)":"petal_length","petal width (cm)":"petal_width"},
             axis=1,inplace=True)
iris.columns

Index(['sepal_length', 'sepal_width', 'petal_length', 'petal_width',
       'species'],
      dtype='object')

In [128]:
iris.head()

sepal_length  sepal_width  petal_length  petal_width  species
0           5.1          3.5           1.4          0.2        0
1           4.9          3.0           1.4          0.2        0
2           4.7          3.2           1.3          0.2        0
3           4.6          3.1           1.5          0.2        0
4           5.0          3.6           1.4          0.2        0

In [129]:
# Cambiar el valor objetivo
iris["species"]=iris.species.map(lambda x: data.target_names[x])
iris.head()

sepal_length  sepal_width  petal_length  petal_width species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa

In [130]:
# Comprobar el valor que falta
iris.isna().sum()

sepal_length    0
sepal_width     0
petal_length    0
petal_width     0
species         0
dtype: int64

Analisis estadistico basico

In [131]:
iris.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   sepal_length  150 non-null    float64
 1   sepal_width   150 non-null    float64
 2   petal_length  150 non-null    float64
 3   petal_width   150 non-null    float64
 4   species       150 non-null    object 
dtypes: float64(4), object(1)
memory usage: 6.0+ KB


In [132]:
iris.describe()

sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.057333      3.758000     1.199333
std        0.828066     0.435866      1.765298     0.762238
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

Analisis de correlacion

In [134]:
# Calcular la correlación solo para las columnas numéricas
iris_numeric = iris.select_dtypes(include=[np.number])
iris_numeric.corr()

sepal_length  sepal_width  petal_length  petal_width
sepal_length      1.000000    -0.117570      0.871754     0.817941
sepal_width      -0.117570     1.000000     -0.428440    -0.366126
petal_length      0.871754    -0.428440      1.000000     0.962865
petal_width       0.817941    -0.366126      0.962865     1.000000

El coeficiente de correlacion de la longitud de los petalos y de la anchura de los petalos es de 0.962865, que es extremadamente alto. Dado que las caracteristicas altamente correlacionadas pueden inducir problemas de multicolinealidad, se recomienda seleccionar una de las dos variables a utilizar.

Analisis de agregacion